### ISSUE TITLE
department-level enable/disable does not work

### ISSUE THREAD

**Issue by jaymzh** [rubocop/rubocop#9752]
**UPDATE**: On further investigation, department-level enable/disable just doesn't work.

The docs [say this should work works](https://docs.rubocop.org/rubocop/1.13/configuration.html):

```
All cops are then disabled by default. Only cops appearing in user configuration files are enabled. Enabled: true does not have to be set for cops in user configuration. They will be enabled anyway. It is also possible to enable entire departments by adding for example

Style:
  Enabled: true
```

But it turns out this doesn't work at all. This is because `Style` is seen as an "invalid" config in Config::Validator.validate() and thus ignored. :/

--------

# Original ticket (for context)

This is probably not a rubocop bug, but I dunno where else to ask.

In the [cookstyle](https://github.com/chef/cookstyle/) project, we just sit ontop of the awesomeness that is Rubocop, and define a bunch of our own Cops and departments.

However, departments defined by our own cops don't seem to be enable-able the way, say `Style` is. So I can do this:

```yaml
AllCops:
  DisabledByDefault: true

Style:
  Enabled: true
```

But what I can't do is (`Chef/Correctness` here is a department, you can see code for cops in that department [here](https://github.com/chef/cookstyle/tree/master/lib/rubocop/cop/chef/correctness)).

```yaml
AllCops:
  DisabledByDefault: true

Chef/Correctness:
  Enabled: true
```

or even:
```yaml
AllCops:
  DisabledByDefault: true

Chef:
  Enabled: true
```

What's really strange is that in robocop I don't see any special setup for the `Style` or `Layout` etc _modules_, so I would think this would "just work"... but it doesn't seem to. When I we try we get:

```
Warning: Chef/Correctness does not support Enabled parameter.

Supported parameters are: 

  - StyleGuideBaseURL
```

If I do just `Chef`, then I get:

```
Error: unrecognized cop Chef found in .cookstyle.yml
```

I'm sure is some minor dumb thing we're doing or not doing that's preventing this from working like the rubocop-built-in departments.

Any help would be greatly appreciated.

**Comment by jaymzh** [comment]
OK I dug through the code, and realized that the code does not work that way:

```
Style:
  Enabled: true
```

is actually just entirely ignored. But the docs [say it works](https://docs.rubocop.org/rubocop/1.13/configuration.html):

```
All cops are then disabled by default. Only cops appearing in user configuration files are enabled. Enabled: true does not have to be set for cops in user configuration. They will be enabled anyway. It is also possible to enable entire departments by adding for example

Style:
  Enabled: true
```

This is because `Style` is seen as an "invalid" config in Config::Validator.validate() and thus ignored. :/

**Comment by jonas054** [comment]
I don't understand. Invalid configuration should make RuboCop stop and report the error, not be ignored. Do you get any error or warning message?

**Comment by jonas054** [comment]
I wrote the above because you said that department level enable/disable doesn't work at all. And that sounds strange to me, because I know that it basically works. :slightly_smiling_face: 

About the original ticket, I can see that `Chef/Correctness does not support Enabled parameter` is reported because there needs to be an `Enabled: true` in the [default configuration of the department](https://github.com/chef/cookstyle/blob/v7.15.1/config/cookstyle.yml#L158). Otherwise it can't be overridden.

Also, there's no concept of department hierarchy implemented in RuboCop, so `Chef` doesn't exist as a department and it cannot be disabled.

I hope that helps.

**Comment by jaymzh** [comment]
OK, so department is always everything up until the last slash, as opposed to everything up to the first slash.

So if you have a rule A/B/C/D you can disable A/B/C (if it has a `Enabled` default) or A/B/C/D, but never A/B or A, is that correct?

I can throw a PR at the chefstyle folks to add a default to `Chef/*`... how hard would it be to extend rubocop to understand department hierarchies?

Thanks!

**Comment by jonas054** [comment]
> OK, so department is always everything up until the last slash, as opposed to everything up to the first slash.

Yes.

> So if you have a rule A/B/C/D you can disable A/B/C (if it has a Enabled default) or A/B/C/D, but never A/B or A, is that correct?

Yes.

> I can throw a PR at the chefstyle folks to add a default to Chef/*... 

Great!

> how hard would it be to extend rubocop to understand department hierarchies?

It's one of those things that feel like it should be fairly easy, but you can't say for sure until you've actually done it. I can try to implement it. Let's keep this issue open for a while.

**Comment by jaymzh** [comment]
Hey @jonas054 since you closed this, should I open up a separate enhancement for hierarchical departments?

**Comment by jonas054** [comment]
Yes, feel free to do so. I'll take a look at the problem. Please tag me in the new issue and reference this issue in it.
